src/de.hpi.swa.graal.squeak/src/de/hpi/swa/graal/squeak/nodes/primitives/impl/ArithmeticPrimitives.java

```diff
@@ -46,12 +47,7 @@ public PrimAddNode(final CompiledMethodObject method) {
             super(method);
         }
 
-        @Specialization(rewriteOn = ArithmeticException.class)
-        protected static final long doLong(final long lhs, final long rhs) {
-            return Math.addExact(lhs, rhs);
-        }
-
-        @Specialization(replaces = "doLong")
+        @Specialization
         protected final Object doLongWithOverflow(final long lhs, final long rhs) {
             return LargeIntegerObject.add(method.image, lhs, rhs);
```

Comment:
Now, that all specializations failing with an `ArithmeticException` are gone, we must make sure that operations such as `add` are not behind a Truffle boundary. Otherwise, we have a performance problem.

Reply:
I think it makes more sense to keep the original specializations that can throw a `ArithmeticException`. As soon as such an exception is thrown, we fall back to the large integer implementation, which retries the long operation, but behind a boundary. The next time, it will go straight to the large integer operation. I think that's good!